uc-crux-llvm/src/UCCrux/LLVM/Run/Check.hs

```diff
@@ -0,0 +1,406 @@
+         -- | Pre-simulation memory
+         MemImpl sym ->
+         -- | Arguments passed to the entry point
+         Assignment (Shape m (SymValue sym arch)) argTypes ->
+         Crux.CruxSimulationResult ->
+         Sim.UCCruxSimulationResult m arch argTypes ->
+         [SomeCheckedCalls m sym arch] ->
+         r) ->
+        r
+    }
+
+data SomeCheckResult m arch =
+  forall argTypes.
+  SomeCheckResult
+    { checkResultFunContext :: FunctionContext m arch argTypes,
+      checkResult :: CheckResult m arch argTypes
+    }
+
+-- NOTE(lb): The explicit kind signature here is necessary for GHC 8.8/8.6
+-- compatibility.
+data TypedConstraints m (argTypes :: Ctx (FullType m))
+  = TypedConstraints
+      { _tcConstraints :: Constraints m argTypes
+      , _tcTypes :: Assignment (FullTypeRepr m) argTypes
+      }
+
+checkInferredContracts ::
+  forall m arch argTypes blocks ret msgs.
+  Crux.Logs msgs =>
+  Crux.SupportsCruxLogMessage msgs =>
+  ArchOk arch =>
+  AppContext ->
+  ModuleContext m arch ->
+  FunctionContext m arch argTypes ->
+  HandleAllocator ->
+  CruxOptions ->
+  LLVMOptions ->
+  Constraints m argTypes ->
+  -- | Entry point
+  Crucible.CFG LLVM blocks (MapToCrucibleType arch argTypes) ret ->
+  -- | Inferred function contracts
+  Map (DefnSymbol m) (Some (TypedConstraints m)) ->
+  IO (CheckResult m arch argTypes)
+checkInferredContracts appCtx modCtx funCtx halloc cruxOpts llOpts constraints cfg contracts =
+   Sim.runSimulatorWithCallbacks
+     appCtx
+     modCtx
+     funCtx
+     halloc
+     constraints
+     cfg
+     cruxOpts
+     llOpts
+     (Sim.SimulatorCallbacks $
+       do ovs <- overrides
+          return $
+            Sim.SimulatorHooks
+              { Sim.createOverrideHooks = map fst ovs
+              , Sim.resultHook =
+                \sym mem args cruxResult ucResult ->
+                  return $
+                    CheckResult $
+                      \k -> k sym mem args cruxResult ucResult (map snd ovs)
+              })
+  where
+    overrides ::
+      IsSymInterface sym =>
+      HasLLVMAnn sym =>
+      IO [ ( Sim.SymCreateOverrideFn sym arch
+           , SomeCheckedCalls m sym arch
+           )
+         ]
+    overrides =
+      for
+        (Map.toList contracts)
+        (\(func, Some (TypedConstraints tcs types)) ->
+           do CFGWithTypes ovCfg argFTys _retTy _varArgs <-
+                pure (findFun modCtx (FuncDefnSymbol func))
+              let argCTys = Crucible.cfgArgTypes ovCfg
+              ovFunCtx <-
+                case makeFunctionContext modCtx func argFTys argCTys of
+                  Left err ->
+                    panic
+                      "checkInferredContracts"
+                      [Text.unpack (ppFunctionContextError err)]
+                  Right funCtx' -> return funCtx'
+              let ?memOpts = CruxLLVM.memOpts llOpts
+              case testEquality argFTys types of
+                Nothing -> panic "checkInferredContracts" []
```

Comment:
Is the error here indicating an unexpected type mismatch?

Reply:
Ah whoops, I meant to go back and add more details here. Yeah, that's exactly right. I'm considering adding another kind of phantom type variable `f` to give type-level names to functions like the current `m` gives type-level names to LLVM modules. Then we could avoid panics/possible problems like this. I'll continue to think about it.